File: packages/rpc/src/RPC/Canonical.ts
/**
 * One value, in a form two of them can be compared as text.
 *
 * Three things in this library have to decide whether two values are *the same value*, and they all
 * arrive at the same question from different directions:
 *
 * - **A row stamp** digests the state a row was read in, so a caller can write only if nothing moved
 *   underneath. See `DataWrites.ts`.
 * - **A projection comparison** decides whether a re-subscribe asks for what the old one asked for,
 *   or for something else. See `sameProjection`, in the client and the server both.
 * - **A cache key** decides whether two `$data` requests are one question. See `@source-repo/query`.
 *
 * They were three encoders, two of them `JSON.stringify`, and that is why this file exists rather
 * than the obvious factoring: `JSON.stringify` answers **key insertion order**, which nothing
 * promises. A JSON column round-trips through a driver, a document store hands back a `BSON` object,
 * and a caller builds an options object in whatever order its code reads - so the same value written
 * twice compares unequal, and each of the three fails differently and quietly. The stamp reports a
 * conflict on a row nobody touched. The projection comparison re-subscribes, spending a targeted
 * snapshot on a slow link to receive what it already had. The cache key misses, and asks the plant
 * again for a page it is holding.
 *
 * One encoder, so the three cannot drift - and the stamp's pinned fixtures in `DataWrites.test.ts`
 * gate all three, because a one-character change here makes every stamp a caller is holding answer
 * *conflict*.
 */

/**
 * The canonical form of one value.
 *
 * **Tagged by kind rather than stringified**, because `1` and `'1'` are different states of a column
 * and an encoding that could not tell them apart would report no change across a type change. Dates
 * and byte arrays are given canonical forms rather than left to `JSON.stringify`, which turns the
 * first into an ISO string and the second into an object of numeric keys - both stable enough by
 * accident today and neither promised by anything.
 *
 * **Object keys are sorted**, which is the one that matters in practice and the reason above.
 *
 * **A key whose value is `undefined` is omitted**, so an absent option and an explicit `undefined`
 * are the same value. That is a claim about all three callers rather than a convenience for one.
 * `{ offset: undefined }` and `{}` describe the same subscription, and treating them as different
 * costs a re-subscribe; `{ filter: undefined }` and `{}` are the same question, and treating them as
 * different costs a round trip on the link this library exists for. It is also the *stamp's* answer:
 * a driver that round-trips a JSON column through JSON drops the key, and one that hands back a live
 * object keeps it as `undefined` - digesting those differently reports a conflict on a row nobody
 * touched, which is the same failure key order produces and is fixed the same way.
 *
 * Total, deliberately: a function or a symbol reaching here is a caller holding it wrong rather than
 * a value a store handed back, and describing it keeps this from throwing inside a precondition.
 */
export const canonicalValue = (value: unknown): unknown => {
    if (value === null || value === undefined) return ['n']
    if (typeof value === 'boolean') return ['b', value]
    if (typeof value === 'number') return ['d', Number.isFinite(value) ? value : String(value)]
    if (typeof value === 'bigint') return ['i', value.toString()]
    if (typeof value === 'string') return ['s', value]
    if (value instanceof Date) return ['t', value.toISOString()]
    if (value instanceof Uint8Array) return ['y', Array.from(value)]
    // An array's order is part of its value - a page of rows in a different order is a different
    // page - so this is the one place order is kept rather than sorted away.
    if (Array.isArray(value)) return ['a', value.map(canonicalValue)]
    if (typeof value === 'object')
        return [
            'o',
            Object.keys(value as Record<string, unknown>)
                .sort()
                .filter((key) => (value as Record<string, unknown>)[key] !== undefined)
                .map((key) => [key, canonicalValue((value as Record<string, unknown>)[key])])
        ]
    return ['?', String(value)]
}

/**
 * The canonical text of one value: what two of them are compared as, and what a digest is taken of.
 *
 * `JSON.stringify` is safe *here* because it is stringifying the canonical form rather than the
 * value - every object in it is an array, so nothing is left for key order to decide.
 */
export const canonicalText = (value: unknown): string => JSON.stringify(canonicalValue(value))

File: packages/csharp/SourceRpc/Canonical.cs
using System.Globalization;
using System.Text;

namespace SourceRpc;

/// <summary>
/// One value, in a form two of them can be compared as text.
///
/// **A port rather than an equivalent**, and the difference matters. This is one of the four things
/// the library specifies once and implements twice - with the outcome rules, the freshness state
/// machine and the deadline arithmetic - and none of them is a fact about the network, so each can
/// differ per language only by being wrong. The reference is `Canonical.ts`, and the fixture that
/// holds the two together is the row stamp's pinned input in `DataWrites.test.ts`, asserted here
/// against the same literal.
///
/// **Tagged by kind rather than stringified**, because `1` and `"1"` are different states of a
/// column and an encoding that could not tell them apart would report no change across a type
/// change. **Object keys are sorted**, because a JSON column round-trips through a driver and a
/// document store hands back BSON, neither of which promises key order between two reads - and
/// digesting insertion order reports a change on a value nobody touched. **A key whose value is
/// absent is omitted**, so an unset option and an explicit null-as-absent are the same value.
///
/// What it is for here, today, is the cache key: two callers asking a peer for the same page with
/// the same filter are asking one question however their options objects were built, and a key that
/// said otherwise would ask the plant twice for a page it is already holding.
/// </summary>
public static class RpcCanonical
{
    /// <summary>
    /// A value that is deliberately absent, as distinct from a value that is null.
    ///
    /// JavaScript has both and .NET has one, which is the single place this port cannot simply
    /// mirror its reference. TypeScript omits an object key whose value is `undefined` and encodes a
    /// `null` as a value; here a property is omitted by being <see cref="Absent"/> and encoded as
    /// null otherwise. Passing `null` where TypeScript would pass `undefined` produces a different
    /// canonical form, so a caller building options for a key should use this for "not asked for".
    /// </summary>
    public static readonly object Absent = new AbsentValue();

    private sealed class AbsentValue
    {
        public override string ToString() => "absent";
    }

    /// <summary>The canonical text of one value: what two of them are compared as.</summary>
    public static string Text(object? value)
    {
        var builder = new StringBuilder();
        Write(builder, value, 0);
        return builder.ToString();
    }

    /// <summary>
    /// A bound, because this may be built from something that arrived over a network. A deeply
    /// nested value is a caller holding it wrong rather than a plant, and recursion without a floor
    /// is a stack overflow rather than an error message.
    /// </summary>
    private const int MaxDepth = 64;

    private static void Write(StringBuilder into, object? value, int depth)
    {
        if (depth > MaxDepth)
        {
            Tagged(into, "?", () => Quote(into, "too deep"));
            return;
        }
        switch (value)
        {
            case null:
            case AbsentValue:
                into.Append("[\"n\"]");
                return;
            case bool flag:
                into.Append(flag ? "[\"b\",true]" : "[\"b\",false]");
                return;
            case string text:
                Tagged(into, "s", () => Quote(into, text));
                return;
            case DateTime moment:
                // Universal and to milliseconds, which is what `Date.prototype.toISOString` writes.
                Tagged(into, "t", () => Quote(into, moment.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss.fffZ", CultureInfo.InvariantCulture)));
                return;
            case DateTimeOffset moment:
                Tagged(into, "t", () => Quote(into, moment.UtcDateTime.ToString("yyyy-MM-ddTHH:mm:ss.fffZ", CultureInfo.InvariantCulture)));
                return;
            case byte[] bytes:
                Tagged(into, "y", () =>
                {
                    into.Append('[');
                    for (var at = 0; at < bytes.Length; at++)
                    {
                        if (at > 0) into.Append(',');
                        into.Append(bytes[at].ToString(CultureInfo.InvariantCulture));
                    }
                    into.Append(']');
                });
                return;
            case System.Numerics.BigInteger big:
                Tagged(into, "i", () => Quote(into, big.ToString(CultureInfo.InvariantCulture)));
                return;
            case sbyte or byte or short or ushort or int or uint or long or ulong or float or double or decimal:
                Tagged(into, "d", () => Number(into, System.Convert.ToDouble(value, CultureInfo.InvariantCulture)));
                return;
            case System.Collections.IDictionary map:
                Tagged(into, "o", () => Fields(into, map.Keys.Cast<object>().Select(key => new KeyValuePair<string, object?>(System.Convert.ToString(key, CultureInfo.InvariantCulture) ?? "", map[key])), depth));
                return;
            case System.Collections.IEnumerable list:
                // An array's order is part of its value - a page of rows in a different order is a
                // different page - so this is the one place order is kept rather than sorted away.
                Tagged(into, "a", () =>
                {
                    into.Append('[');
                    var first = true;
                    foreach (var item in list)
                    {
                        if (!first) into.Append(',');
                        first = false;
                        Write(into, item, depth + 1);
                    }
                    into.Append(']');
                });
                return;
            default:
                Tagged(into, "o", () => Fields(into, Properties(value), depth));
                return;
        }
    }

    /// <summary>
    /// A plain object's readable properties, which is what an options record is in .NET.
    ///
    /// Public instance properties with a getter and no index, which covers a record, an anonymous
    /// type and a POCO alike - and deliberately not fields, so that a private backing store cannot
    /// reach a key twice under two names.
    /// </summary>
    private static IEnumerable<KeyValuePair<string, object?>> Properties(object value) =>
        value.GetType()
            .GetProperties(System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Instance)
            .Where(property => property.CanRead && property.GetIndexParameters().Length == 0)
            .Select(property => new KeyValuePair<string, object?>(property.Name, property.GetValue(value)));

    private static void Fields(StringBuilder into, IEnumerable<KeyValuePair<string, object?>> fields, int depth)
    {
        into.Append('[');
        var first = true;
        // Ordinal, which is what `Array.prototype.sort` does to strings by default - a culture-aware
        // comparison would order the same two keys differently on two machines, which is the failure
        // sorting exists to prevent rather than a new one.
        foreach (var field in fields.Where(field => field.Value is not AbsentValue).OrderBy(field => field.Key, StringComparer.Ordinal))
        {
            if (!first) into.Append(',');
            first = false;
            into.Append('[');
            Quote(into, field.Key);
            into.Append(',');
            Write(into, field.Value, depth + 1);
            into.Append(']');
        }
        into.Append(']');
    }

    private static void Tagged(StringBuilder into, string tag, Action body)
    {
        into.Append("[\"").Append(tag).Append("\",");
        body();
        into.Append(']');
    }

    /// <summary>
    /// A number as `JSON.stringify` writes it.
    ///
    /// The one place this port has to imitate a runtime rather than a specification, so what it
    /// covers is stated rather than assumed. An integral value below 2^53 is written without a
    /// decimal point, which is what JavaScript does and what .NET does not; everything else uses
    /// the shortest round-trip form, which both runtimes agree on for ordinary decimals. **Very
    /// large and very small magnitudes are where the two could still diverge**, because the
    /// threshold at which each switches to exponent notation is a runtime's choice - a value out
    /// there in a cache key is harmless, and one in a row stamp would be a genuine disagreement.
    /// Non-finite values never reach here; they are strings by then, since a sensor reading
    /// out-of-range and a sensor reading no-value must not become the same state.
    /// </summary>
    private static void Number(StringBuilder into, double value)
    {
        if (double.IsNaN(value) || double.IsInfinity(value))
        {
            Quote(into, double.IsNaN(value) ? "NaN" : value > 0 ? "Infinity" : "-Infinity");
            return;
        }
        if (value == 0)
        {
            // Negative zero prints as "0" in JavaScript, and two readings of zero are one state.
            into.Append('0');
            return;
        }
        if (Math.Floor(value) == value && Math.Abs(value) < 9007199254740992d)
        {
            into.Append(((long)value).ToString(CultureInfo.InvariantCulture));
            return;
        }
        into.Append(value.ToString("R", CultureInfo.InvariantCulture).Replace("E", "e", StringComparison.Ordinal));
    }

    /// <summary>
    /// A string as `JSON.stringify` writes it: quote and backslash escaped, the five short escapes
    /// used where they exist, every other control character as `\uXXXX`, and everything above them
    /// - including non-ASCII - left exactly as it is.
    ///
    /// Written out rather than delegated, because .NET's JSON writer escapes more than JavaScript
    /// does by default: `&lt;`, `&gt;`, `&amp;` and non-ASCII all come back as `\uXXXX`, which is
    /// valid JSON and a different string. A tag name with an umlaut in it would then hash
    /// differently in the two languages, which is exactly the divergence this file exists to prevent.
    /// </summary>
    private static void Quote(StringBuilder into, string text)
    {
        into.Append('"');
        foreach (var character in text)
        {
            switch (character)
            {
                case '"': into.Append("\\\""); break;
                case '\\': into.Append("\\\\"); break;
                case '\b': into.Append("\\b"); break;
                case '\f': into.Append("\\f"); break;
                case '\n': into.Append("\\n"); break;
                case '\r': into.Append("\\r"); break;
                case '\t': into.Append("\\t"); break;
                default:
                    if (character < 0x20) into.Append("\\u").Append(((int)character).ToString("x4", CultureInfo.InvariantCulture));
                    else into.Append(character);
                    break;
            }
        }
        into.Append('"');
    }
}
